File: clients/csharp/Model/JumpState.cs
namespace AiCup2019.Model
{
    public struct JumpState
    {
        public bool CanJump { get; set; }
        public double Speed { get; set; }
        public double MaxTime { get; set; }
        public bool CanCancel { get; set; }
        public JumpState(bool canJump, double speed, double maxTime, bool canCancel)
        {
            this.CanJump = canJump;
            this.Speed = speed;
            this.MaxTime = maxTime;
            this.CanCancel = canCancel;
        }
        public static JumpState ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new JumpState();
            result.CanJump = reader.ReadBoolean();
            result.Speed = reader.ReadDouble();
            result.MaxTime = reader.ReadDouble();
            result.CanCancel = reader.ReadBoolean();
            return result;
        }
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(CanJump);
            writer.Write(Speed);
            writer.Write(MaxTime);
            writer.Write(CanCancel);
        }
    }
}

File: clients/fsharp/Model/JumpState.fs
namespace AiCup2019.Model

type JumpState =
    {
        CanJump: bool
        Speed: double
        MaxTime: double
        CanCancel: bool
    } with
    member this.writeTo (writer: System.IO.BinaryWriter) =
        writer.Write this.CanJump
        writer.Write this.Speed
        writer.Write this.MaxTime
        writer.Write this.CanCancel

    static member readFrom (reader: System.IO.BinaryReader) =
        {
            CanJump = reader.ReadBoolean()
            Speed = reader.ReadDouble()
            MaxTime = reader.ReadDouble()
            CanCancel = reader.ReadBoolean()
        }
